src/languages/javascript.js

```diff
@@ -63,12 +63,26 @@ export default function(hljs) {
     literal: ECMAScript.LITERALS.join(" "),
     built_in: ECMAScript.BUILT_INS.join(" ")
   };
+  const nonDecimalLiterals = (prefixLetters, validChars) => 
+    `\\b0[${prefixLetters}][${validChars}]([${validChars}_]*[${validChars}])?n?`;
+  const noLeadingZeroDecimalDigits = /[1-9]([0-9_]*\d)?/;
+  const decimalDigits = /\d([0-9_]*\d)?/;
+  const exponentPart = regex.concat(/[eE][+-]?/, decimalDigits);
   const NUMBER = {
     className: 'number',
     variants: [
-      { begin: '\\b(0[bB][01]+)n?' },
-      { begin: '\\b(0[oO][0-7]+)n?' },
-      { begin: hljs.C_NUMBER_RE + 'n?' }
+      { begin: nonDecimalLiterals('bB', '01') }, // Binary literals
+      { begin: nonDecimalLiterals('oO', '0-7') }, // Octal literals
+      { begin: nonDecimalLiterals('xX', '0-9a-fA-F') }, // Hexadecimal literals
+      { begin: regex.concat(/\b/, noLeadingZeroDecimalDigits, 'n') }, // Non-zero BigInt literals
+      { begin: regex.concat(/0?\./, decimalDigits, regex.optional(exponentPart)) }, // Decimal literals between 0 and 1
+      { begin: regex.concat(
+        /\b/,
+        noLeadingZeroDecimalDigits,
+        regex.optional(regex.concat(/\./, regex.optional(decimalDigits))), // fractional part
+        regex.optional(exponentPart)
+        ) }, // Decimal literals >= 1
+      { begin: /\b0[\.n]?/ }, // Zero literals (`0`, `0.`, `0n`)
```

Comment:
Is sequencing off here?  Isn't it `[.]?n?`?  `0n.` would not be correct, would it?

Reply:
Oh, it's either... :-) So then I think this is ok since `.0n` does not seem valid.